messaging/messaging.go

```diff
@@ -0,0 +1,460 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package messaging contains functions for sending messages and managing
+// device subscriptions with Firebase Cloud Messaging (FCM).
+package messaging
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"firebase.google.com/go/internal"
+	"google.golang.org/api/transport"
+)
+
+const (
+	messagingEndpoint = "https://fcm.googleapis.com/v1"
+	iidEndpoint       = "https://iid.googleapis.com"
+	iidSubscribe      = "iid/v1:batchAdd"
+	iidUnsubscribe    = "iid/v1:batchRemove"
+)
+
+var (
+	topicNamePattern = regexp.MustCompile("^(/topics/)?(private/)?[a-zA-Z0-9-_.~%]+$")
+
+	fcmErrorCodes = map[string]string{
+		"INVALID_ARGUMENT":   "request contains an invalid argument; code: invalid-argument",
+		"NOT_FOUND":          "request contains an invalid argument; code: registration-token-not-registered",
+		"PERMISSION_DENIED":  "client does not have permission to perform the requested operation; code: authentication-error",
+		"RESOURCE_EXHAUSTED": "messaging service quota exceeded; code: message-rate-exceeded",
+		"UNAUTHENTICATED":    "client failed to authenticate; code: authentication-error",
+		"UNAVAILABLE":        "backend servers are temporarily unavailable; code: server-unavailable",
+	}
+
+	iidErrorCodes = map[string]string{
+		"INVALID_ARGUMENT": "request contains an invalid argument; code: invalid-argument",
+		"NOT_FOUND":        "request contains an invalid argument; code: registration-token-not-registered",
+		"INTERNAL":         "server encountered an internal error; code: internal-error",
+		"TOO_MANY_TOPICS":  "client exceeded the number of allowed topics; code: too-many-topics",
```

Comment:
Oh no, iid can have a non-standard error value? Hmm...

Reply:
It is documented here: https://developers.google.com/instance-id/reference/server#manage_relationship_maps_for_multiple_app_instances